### ISSUE TITLE
hierarchical directed layout: edge crossing

### ISSUE THREAD

**Issue by UnicornDeluxe** [visjs/vis-network#84]
The edge crossing bug seems to be [known](https://github.com/almende/vis/issues/3357) but it still would be nice if it gets fixed or at least if one could toggle if he want "optimazation but crossing" or "no crossing but no optimization" mentioned [here](https://github.com/almende/vis/issues/3357#issuecomment-322148978).

Maybe this is to some degree related to https://github.com/visjs/vis-network/issues/83.

**Comment by Thomaash** [comment]
Is it always possible to organize a hierarchy without any edges crossing? I think there will be cases where it's simply impossible in 2D space.

**Comment by UnicornDeluxe** [comment]
@Thomaash I do agree, in case of **cross references** between nodes, edges will have to cross each other. So there is not always a possibilty to organize a hierarchy without any edges crossing. 

In some cases crossing can be avoided if "no edge crossing" is prioritised over "keeping the configured space between nodes".  (first graph of this fiddle shows what i mean http://jsfiddle.net/huwp23bz/)
This fiddle http://jsfiddle.net/m56qtkcu/ also shows "edge with node crossing" where on the top graph, node 3 overlaps with edge from 1 to  5.

**Comment by dawidchodura** [comment]
I try to render a tree with hierarchical directed layout, but the edges get crossed. Would it help if it was possible to specify in the options that the graph is a tree (ie. no cross references), so that the algorithm could take a hint, while laying out the nodes?

**Comment by Thomaash** [comment]
Hi @UnicornDeluxe and @dawidchodura,

I was thinking about this and I think we should add a new option which could look like this:
```typescript
interface HorizontalLayout {
  algorithm?:
    // the way it is now and would be the default
    | 'none'
    // fast O(N) and perfect for trees, will fail on multiple paths to the same node
    | 'tree'
    // some heuristics to get a reasonably good layout for a generic graph
    // no idea how to implement this ATM, maybe not a part of the first PR
    | 'best-effort'
  propNameOrGetter?:
    // name of the property that will be used for ordering the nodes
    // would default to id (just to make it stable as ids generally don't change)
    | string
    // node comparator (like Array.sort callback for given nodes descendants)
    | ((a: NodeOptions, b: NodeOptions) => number)
}
```

Do you think this would satisfy your needs?
Do you have any objections or improvements?

PS: If you can come up with better more descriptive names for the properties or string literals I'm all ears.
PPS: @mojoaxel do you want to join this discussion?

**Comment by UnicornDeluxe** [comment]
Looks good to me.

Concerning the "best-effort" algorithm, I guess the _Reduction of the Number of Edge Crossings_ part of the [Sugiyama-Algorithm](https://www.researchgate.net/profile/Nikola_Nikolov2/publication/303226437_Sugiyama_Algorithm/links/59354e6f45851553b6f39e2e/Sugiyama-Algorithm.pdf) can help.

**Comment by deepanshu11madan** [comment]
@UnicornDeluxe were you able to achieve no edge crossing? i looked into Sugiyama, and Dagre layout implements it so i am using dagre to calculate coordinates for nodes and edges. With visjs i can set the nodes at those coordinates determined by dagre but no way to tell vsjs edges point coordinates

**Comment by UnicornDeluxe** [comment]
@deepanshu11madan no, I wasn't able to achieve no edge crossing. What do you mean by
> no way to tell vsjs edges point coordinates

?
Can you share a screenshot of the resulting graph?

**Comment by deepanshu11madan** [comment]
So dagre is a layered graph layout algorithm based on sugiyama, it calculates coordinates for nodes and control points for edges. With Visjs I can place the nodes on the coordinates generated by dagre, which works perfectly, but the edges are rendered by Visjs and they cross and overlap. I have the control points for edges generated by the dagre algorithm, but no way to tell visjs to use those array of points to draw edges instead. Does that make sense?

I do have a enhancement issue created which Thomaash plans to work on at some point. Meanwhile I was wondering if you were able to figure the edge crossing issue.

**Comment by UnicornDeluxe** [comment]
That does indeed make sense.
Have you tried placing 'fake nodes' at the mentioned control points for the edges to force visjs to draw the edges as you want?
So for example:
![grafik](https://user-images.githubusercontent.com/43786929/70848889-0363c380-1e78-11ea-844c-2f45cbc1649c.png)
Dagre gives you the coordinates of nodes `a, b, c and d` and a control point `e` for the edge from `b to d` you can simply set a node at `e` and split the edge from `b to d` into `b to e` and `e to d`. Make the node `e` invisible and you kind of told vis to use this specific point to draw the edge from `b to d`.
Which is what you/we want. Does this work?

**Comment by deepanshu11madan** [comment]
yeah actually that's I am doing right now, but i have edge properities like  dragging edges, edge, arrows, labels which doesn't work well, will need to dirty hacks and the corners are also not very smooth, for example , in the image below, do you know if it can be made smooth somehow, the connection point of dummy node
![Screen Shot 2019-12-14 at 3 44 53 PM](https://user-images.githubusercontent.com/47491495/70856006-061ae480-1e89-11ea-97db-5457bf998f62.png)

**Comment by UnicornDeluxe** [comment]
you can try to reduce the size of dummy nodes, but otherwise I have no idea

**Comment by wyleyr** [comment]
Wondering if any progress has been made on this issue. I am running into edge crossing problems when using vis to visualize WordNet networks. I don't have much new information to add to this issue, but could provide some further examples of networks where edge crossing occurs even though the graph can be laid out without edge crossing. Would that be helpful?

**Comment by Jmedders** [comment]
I've been utilizing this example https://visjs.github.io/vis-network/examples/network/other/configuration.html to easily recreate this issue. I've tried playing with all of the options to no avail. To add some visuals to this issue, here is an image of the result when to arrows are added, layout is set to hierarchal, sortMethod is directed, and shakeTowards: roots

<img width="595" alt="Screen Shot 2020-04-30 at 5 09 44 PM" src="https://user-images.githubusercontent.com/16631297/80767667-a9515b00-8b05-11ea-9e57-462e4c60cb88.png">
By moving the nodes around manually was able to produce the ideal solution, but it'd be amazing if there was a way to achieve something similar to this without the manual process of moving nodes around
<img width="601" alt="Screen Shot 2020-04-30 at 5 08 26 PM" src="https://user-images.githubusercontent.com/16631297/80767690-bd955800-8b05-11ea-8ee9-c7c7f5620cc9.png">

**Comment by whme** [comment]
Hi @Thomaash, I was wondering if you have made any progress concerning this issue 🙃

**Comment by Thomaash** [comment]
Hi @whme, unfortunately no.

**Comment by Xerusial** [comment]
The problem is, that edge pulling force is less than node repulsion. What if node repulsion is disabled till the network is stable for the first time and edges have minimal length and then node repulsion kicks in and seperates the nodes from eachother?

**Comment by wyleyr** [comment]
@Xerusial can you elaborate on this? How did you determine that that was the problem? Do you have some sense of what would be required to code up the solution you propose?

**Comment by Thomaash** [comment]
Hi @Xerusial,

your solution makes no sense to me but, as always, that doesn't mean it's wrong (I don't understand everything). Could you open a PR so that we can discuss the code? I don't have the time to fix this myself right now but I will definitely find some time to review and hopefully merge a solution.

Thanks.

**Comment by pascalbetz** [comment]
Hi

I'm struggling with this as well.
Is there a way to influence the ordering on the cross axis? Assuming I have a "UD" hierarchical layout, I could solve my "edges crossing" problem by specifying not only the _level_ (Y-Axis) but also the _cross-axis-level_ (X-Axis) (or vice verca for LR/RL direction)

**Comment by DurkoMatko** [comment]
Havinf the same issue...with sortMethod: "directed", edges get crossed on initial render. Manually, I'm able to create a graph with 0 edge crosses while preventing the hierarchy

**Comment by Ronnocerman** [comment]
@Thomaash

I'm dealing with this too.

I think @Xerusial is basically pointing out that the spring force can't overcome the repulsive force to un-cross the nodes in a hierarchical layout. Is it possible for nodes to "pull past" each other in a hierarchical layout if the spring force of the "parent" nodes is greater than the repulsive force between children?

**Comment by rolffank** [comment]
@Thomaash

> Is it always possible to organize a hierarchy without any edges crossing? I think there will be cases where it's simply impossible in 2D space.

You can test if the graph is planar. If so, it is possible to draw it without edge crossings (but I don't know if a hierarchical layout is always possible in planar cases).
The dot layout engine of Graphviz minimizes edge crossings very well. I think they use Sugiyama algorithm. I used it to get the coordinates of the nodes because in viz.js (JavaScript version of Graphviz created with emscripten) it's not possible to move the nodes manually. Cytoscapejs also use dagre's layout.

**Comment by dimitriz09** [comment]
Subscribing to this as it can truly break the reading experience. 
This 3 nodes graph is rendered flat. 

![Screenshot from 2024-03-23 14-16-53](https://github.com/visjs/vis-network/assets/109512997/2e63afe3-d4e8-4908-9ef5-60b5c5694be1)


https://jsfiddle.net/uadc4n05/
